Set GlobalOpenTelemetry to noop

Yas-Franco opened the pull request:
For issue #23923 

- GlobalOpenTelmetry is always returned as OpenTelemetry.noop()
- Users are warned when getting the GlobalOpenTelemetry object
- Users are unable to set the GlobalOpenTelemetry object themselves

Yas-Franco committed:
Set GlobalOpenTelemetry to noop

Yas-Franco commented:
#libby
#build

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryGlobalOpenTelemetryTest.java:
```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
```
2023

Yas-Franco replied:
Done!

Azquelt reviewed (commented):
The code here looks good :+1:

We have an odd merge commit, merging integration into this branch, is that necessary?

We also want to output a warning if the user tries to use the `GlobalOpenTelemetry` object.

Yas-Franco commented:
I have updated the copyright, removed the unnecessary commit and fixed the bundle imports that was causing test failures. A warning message still needs adding

#libby
#build

Yas-Franco commented:
#build
#libby

Azquelt commented on dev/io.openliberty.io.opentelemetry.internal/src/io/opentelemetry/api/GlobalOpenTelemetry.java:
```diff
@@ -0,0 +1,64 @@
+    private static final TraceComponent tc = Tr.register(GlobalOpenTelemetry.class);
+    private static boolean warned = false;
+
+    public static OpenTelemetry get() {
+        //Warns the user the first time they try to get the GlobalOpenTelemetry instance
+        //The GlobalOpenTelemetry instance should not be used as the entrypoint to telemetry functionality.
+        //OpenTCWMOT5000W: Cannot get GlobalOpenTelemetry.get
+        if (!warned) {
+            String msg = Tr.formatMessage(tc, "CWMOT5000.cannot.get.globalopentelemetry");
+            Tr.warning(tc, msg);
+            warned = true;
+        }
+        return OpenTelemetry.noop();
+    }
```
I remembered that we use `GlobalOpenTelemetry.get()` to get the `OpenTelemetry` instance when using the agent, so we may see the Agent tests fail.

They may not fail, I'm not sure how the classloaders work when the agent is attached and which version of the class will be visible to our code. This version may be hidden completely by the agent in which case it would work.

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryGlobalOpenTelemetryTest.java:
```diff
@@ -0,0 +1,78 @@
+    public void testGetGlobalOpenTelemetry() throws Exception {
+        server.setMarkToEndOfLog();
+        runTest(server, APP_NAME + "/GlobalOpenTelemetryServlet", "testGetGlobalOpenTelemetry");
+        assertFalse(server.waitForStringInLogUsingMark("CWMOT5000W: Cannot get GlobalOpenTelemetry.").isEmpty());
```
`waitForStringInLogUsingMark` returns `null` if the string is not found. This line will throw a `NullPointerException` if the warning isn't emitted.

Azquelt reviewed (approved):
I've left a couple of comments but I think this can be merged as-is as long as the tests pass and any tidying up done later.

Obviously if the tests fail, we need to rethink the approach.

Azquelt commented:
This also needs L2 and ID review before it can be merged.

Azquelt commented on dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops:
```diff
@@ -0,0 +1,34 @@
+# Message prefix block: CWMOT5000 - CWMOT5999
+#-----------------------------------------------------------------------------------------------------------------------------
+
+CWMOT5000.cannot.get.globalopentelemetry=CWMOT5000W: Cannot get GlobalOpenTelemetry.
+CWMOT5000.cannot.get.globalopentelemetry.explanation=The GlobalOpenTelemetry instance should not be used as the entrypoint to telemetry functionality.
+CWMOT5000.cannot.get.globalopentelemetry.useraction=Use the OpenTelemetry instance instead.
```
Ok, I didn't reveiw this properly, I think we need a bit more explanation here:

I think we can include in the message what has happened and what the user should do instead.

```suggestion
CWMOT5000.cannot.get.globalopentelemetry=CWMOT5000W: The GlobalOpenTelemetry.get method was called. This method returns a non-functional OpenTelemetry object. Use CDI to inject an OpenTelemetry object instead.
CWMOT5000.cannot.get.globalopentelemetry.explanation=There is no global OpenTelemetry object and so using GlobalOpenTelemetry is not supported. Each application must instead use CDI to inject the OpenTelemetry object for that application.
CWMOT5000.cannot.get.globalopentelemetry.useraction=Use CDI to inject the OpenTelemetry object.
```

Azquelt commented on dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWMOT5000.cannot.get.globalopentelemetry.explanation=The GlobalOpenTelemetry instance should not be used as the entrypoint to telemetry functionality.
+CWMOT5000.cannot.get.globalopentelemetry.useraction=Use the OpenTelemetry instance instead.
+
+CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Cannot set GlobalOpenTelemetry.
+CWMOT5001.cannot.set.globalopentelemetry.explanation=The GlobalOpenTelemetry instance should not be used as the entrypoint to telemetry functionality.
+CWMOT5001.cannot.set.globalopentelemetry.useraction=Use the OpenTelemetry instance instead.
```
```suggestion
CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Setting GlobalOpenTelemetry is not supported
CWMOT5001.cannot.set.globalopentelemetry.explanation=There is no global OpenTelemetry object and so calling the GlobalOpenTelemetry.set method is not supported. Each application must instead use CDI to inject the OpenTelemetry object for that application.
CWMOT5001.cannot.set.globalopentelemetry.useraction=Do not call the GlobalOpenTelemetry.get method. Where the application code needs to obtain an OpenTelemetry object, use CDI to inject it.
```

Yas-Franco committed:
Update dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops

Co-authored-by: Andrew Rouse <anrouse@uk.ibm.com>

Yas-Franco committed:
Update dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops

Co-authored-by: Andrew Rouse <anrouse@uk.ibm.com>

Yas-Franco commented:
#build
#libby

posmith commented:
L2 message review completed.

Yas-Franco committed:
Update globalopentelemetry test messages

Yas-Franco commented:
#build
#libby

Azquelt commented:
This is ready to merge as soon as it's been reviewed by ID @OpenLiberty/message-reviewer

dmuelle commented on dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops:
```diff
@@ -0,0 +1,34 @@
+#-----------------------------------------------------------------------------------------------------------------------------
+
+CWMOT5000.cannot.get.globalopentelemetry=CWMOT5000W: The GlobalOpenTelemetry.get method was called. This method returns a non-functional OpenTelemetry object. Use CDI to inject an OpenTelemetry object instead.
+CWMOT5000.cannot.get.globalopentelemetry.explanation=There is no global OpenTelemetry object and so using GlobalOpenTelemetry is not supported. Each application must instead use CDI to inject the OpenTelemetry object for that application.
```
```suggestion
CWMOT5000.cannot.get.globalopentelemetry.explanation= The GlobalOpenTelemetry class is not supported because no global OpenTelemetry object exists. Instead, each application must use CDI to inject the OpenTelemetry object for that application.
```

dmuelle replied:
Avoid using "There is..." construction as it obscures subject of sentence and can cause trouble for translators

dmuelle commented on dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWMOT5000.cannot.get.globalopentelemetry.useraction=Use CDI to inject the OpenTelemetry object.
+
+CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Setting GlobalOpenTelemetry is not supported
+CWMOT5001.cannot.set.globalopentelemetry.explanation=There is no global OpenTelemetry object and so calling the GlobalOpenTelemetry.set method is not supported. Each application must instead use CDI to inject the OpenTelemetry object for that application.
```
```suggestion
CWMOT5001.cannot.set.globalopentelemetry.explanation= The GlobalOpenTelemetry class is not supported because no global OpenTelemetry object exists. Instead, each application must use CDI to inject the OpenTelemetry object for that application.
```

dmuelle commented on dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWMOT5000.cannot.get.globalopentelemetry.explanation=There is no global OpenTelemetry object and so using GlobalOpenTelemetry is not supported. Each application must instead use CDI to inject the OpenTelemetry object for that application.
+CWMOT5000.cannot.get.globalopentelemetry.useraction=Use CDI to inject the OpenTelemetry object.
+
+CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Setting GlobalOpenTelemetry is not supported
```
```suggestion
CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Setting the GlobalOpenTelemetry class is not supported.
```

dmuelle reviewed (changes requested):
ID review- let me know if you have any questions or if I've misinterpreted anything. Thanks

Yas-Franco committed:
Correct GlobalOpenTelemetry warning messages

Yas-Franco commented:
Thanks @dmuelle  - The changes are up 
 
 Rebased with [yasmin-aumeeruddy:telemetry-warning-tracing](https://github.com/OpenLiberty/open-liberty/pull/24397) to prevent merge conflicts 
 
#libby
#build

dmuelle reviewed (dismissed):
LGTM thanks

Azquelt commented on dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWMOT5000.cannot.get.globalopentelemetry.explanation=The GlobalOpenTelemetry class is not supported because no global OpenTelemetry object exists. Instead, each application must use CDI to inject the OpenTelemetry object for that application.
+CWMOT5000.cannot.get.globalopentelemetry.useraction=Use CDI to inject the OpenTelemetry object.
+
+CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Setting the GlobalOpenTelemetry class is not supported.
```
I agree with the restructuring suggested by ID, but we should have "object" rather than "class" here.
```suggestion
CWMOT5001.cannot.set.globalopentelemetry=CWMOT5001E: Setting the GlobalOpenTelemetry object is not supported.
```

Yas-Franco committed:
Update dev/io.openliberty.io.opentelemetry.internal/resources/io/openliberty/io/opentelemetry/resources/MPTelemetry.nlsprops

Co-authored-by: Andrew Rouse <anrouse@uk.ibm.com>

Yas-Franco committed:
Update TelemetryGlobalOpenTelemetryServlet.java

Yas-Franco commented:
#build
#libby
